website/server/libs/auth.js

```diff
@@ -0,0 +1,41 @@
+import validator from 'validator';
+import * as passwordUtils from './password';
+import { async } from 'async';
+import {
+  model as User,
+} from '../models/user';
+import {
+  NotAuthorized
+} from './errors';
+
+module.exports.checkCredentials = function (username, password, done) {
+    let login;
+
+    if (validator.isEmail(username)) {
+      login = {'auth.local.email': username.toLowerCase()}; // Emails are stored lowercase
+    } else {
+      login = {'auth.local.username': username};
+    }
+
+    // load the entire user because we may have to save it to convert the password to bcrypt
+    User.findOne(login).exec().then((user)=>{

```

Comment:
Two advices for Promises:

1. You can return the promise from this method so that we can choose to use either the promise or the callback to deal with results and errors